src/logging/metrics-logger.ts

```diff
@@ -0,0 +1,128 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {hrtime} from 'process';
+import {Event} from '../event.js';
+import {DefaultLogger} from './default-logger.js';
+
+interface Metric {
+  name: string;
+  matches: (event: Event) => boolean;
+  count: number;
+}
+
+/**
+ * A {@link Logger} that keeps track of metrics.
+ */
+export class MetricsLogger extends DefaultLogger {
+  private _startTime: [number, number] = hrtime();
+  private readonly _metrics: Metric[] = [
+    {
+      name: 'Success',
+      // 'no-command' is technically a success, but we don't want to count it as
+      // a success for this metric because nothing was actually run.
+      matches: (e: Event) => e.type === 'success' && e.reason !== 'no-command',
+      count: 0,
+    },
+    {
+      name: 'Ran',
+      matches: (e: Event) => e.type === 'success' && e.reason === 'exit-zero',
+      count: 0,
+    },
+    {
+      name: 'Skipped (fresh)',
+      matches: (e: Event) => e.type === 'success' && e.reason === 'fresh',
+      count: 0,
+    },
+    {
+      name: 'Restored from cache',
+      matches: (e: Event) => e.type === 'success' && e.reason === 'cached',
+      count: 0,
+    },
+  ];
+
+  /**
+   * @param rootPackage The npm package directory that the root script being
+   * executed belongs to.
+   */
+  constructor(rootPackage: string) {
+    super(rootPackage);
+  }
+
+  /**
+   * Update relevant metrics for an event and pass it up to the parent logger.
+   */
+  override log(event: Event): void {
+    // When in watch mode, metrics should reset at the start of each run.
+    if (event.type === 'info' && event.detail === 'watch-run-start') {
```

Comment:
Not the biggest deal, but it seems a little asymmetrical that this class intrinsically understands when it should reset its metrics, but requires the caller to tell it when to display the metrics.

One idea I had was to make each instance of a `MetricsLogger` only live for the duration of a single iteration. So, in one-shot mode there would be one instance, but in watch mode we would create a new one on each iteration. The benefit of that is that we don't need any reset logic -- we can just use class field initializers to set the counters to 0 and the start time to now.

I'm also fine with it as-is, though.

Reply:
I think this is a good suggestion, looking into it. I explored this a bit when I got started with this work, but couldn't identify an existing event that signals 'the last service has finished executing', which I assume is when we would want MetricsLogger to automatically print metrics to the console. 

Do you know if there is an existing way to know when all services have finished executing, or would we need a new event (or some other signal)?

Reply:
Yeah, maybe it's not so straightforward as I imagined. This is great as-is, we can always refactor things later if needed.